### PR TITLE
SSL Fine-tuning PR

### PR DESCRIPTION
**Pull request by wanchichen**
## What?

This PR adds support for fine-tuning SSL models pre-trained in ESPnet. 
It supports two different interfaces for downstream use:

1. Initialize manually from params. Better for task-specific custom training code (inter-ctc)
        
```
   encoder: transformer
   encoder_conf:
       output_size: 768
   init_param: ['hubert.pt:encoder:encoder']
```

 2. Use wrapper, to use the SSL model's original forward function or S3PRL style
        

   ```
      input_size: none
      frontend: espnet_ssl
      frontend_conf:
          frontend_conf:
              path_or_url: hubert.pt
          freeze_encoder_steps: 6000
          multilayer_feature: False
          mask_feats: True
      normalize: null
      specaug: null
      encoder: null
   ```

For now i only did ASR, but maybe I should just do all tasks that use SSL a lot (ST, SLU)?

### CODE REVIEW

**Comment by sw005320** [comment]
It's better to increase the time for the test.

File: espnet2/asr/espnet_model.py [hunk 1]
```diff
@@ -103,9 +112,13 @@ def __init__(
         self.postencoder = postencoder
         self.encoder = encoder
 
-        if not hasattr(self.encoder, "interctc_use_conditioning"):
+        self.autocast_frontend = autocast_frontend
+
+        if self.encoder is not None and not hasattr(
```
Comment (simpleoier) [inline espnet2/asr/espnet_model.py:117]: Would it be more readable to write it as 
```
if self.encoder and getattr(self.encoder, "interctc_use_conditioning", False):
    self.encoder.conditioning_layer = ...
```

File: espnet2/asr/frontend/cnn.py [hunk 2]
```diff
@@ -131,6 +173,7 @@ def __init__(
         fs: Union[int, str] = 16000,
         normalize_audio: bool = False,
         normalize_output: bool = False,
+        layer_norm_cls: str = "transposed",
```
Comment (simpleoier) [inline espnet2/asr/frontend/cnn.py:176]: How about using `layer_norm_cls: typing.Literal["transposed", "dim1"] = "transposed"`?

File: espnet2/tasks/asr.py [hunk 3]
```diff
@@ -568,12 +572,16 @@ def build_model(cls, args: argparse.Namespace) -> ESPnetASRModel:
...
 
         # 4. Encoder
-        encoder_class = encoder_choices.get_class(args.encoder)
-        encoder = encoder_class(input_size=input_size, **args.encoder_conf)
+        if getattr(args, "encoder", None) is not None:
+            encoder_class = encoder_choices.get_class(args.encoder)
+            encoder = encoder_class(input_size=input_size, **args.encoder_conf)
+            encoder_output_size = encoder.output_size()
+        else:
+            encoder = None
```
Comment (simpleoier) [inline espnet2/tasks/asr.py:580]: I think a test for this change is necessary and important.

**Review by simpleoier — APPROVED** [review]
Thanks! This PR looks good to me.

**Comment by wanchichen** [comment]
> Thanks! This PR looks good to me.

Thanks a lot Xuankai! I'll add the recommended fixes.

**Comment by sw005320** [comment]
@wanchichen, what is the status of this PR?

**Comment by wanchichen** [comment]
ah the status is that I completely forgot about it. restarted by addressing some of xuankai's comments, just need to add the last unit test now

**Comment by sw005320** [comment]
This pull request introduces significant updates to the ESPnet ASR framework, focusing on new training configurations, enhancements to convolutional positional embeddings, and support for flexible encoder configurations. These changes aim to improve model training, increase flexibility in architecture, and enhance computational efficiency.

### Training Configuration Updates:
* Added three new training configuration files: `train_espnet_ssl_hubert_frontend.yaml`, `train_espnet_ssl_hubert_init.yaml`, and `train_espnet_ssl_hubert_multilayer.yaml`. These configurations introduce support for different frontends (`espnet_ssl`, `wav2vec_cnn`), initialization methods, and encoder settings, along with features like multi-layer support and SpecAugment. [[1]](diffhunk://#diff-174bcdd3d578d05c84a9fca8bb37ba09d103f05f4635d000743ec19b432280aeR1-R43) [[2]](diffhunk://#diff-71ee90c611fdc38195db3114196228ef9323ac260c034ac2c5350ce091b934bfR1-R83) [[3]](diffhunk://#diff-00c5802997d1e2652d9cff7bab7e245609f6ced00b00be837809a034b942ab14R1-R56)

### Enhancements to Convolutional Positional Embeddings:
* Added new parameters (`num_layers`, `weight_norm`, `use_residual`) to the `ConvolutionalPositionalEmbedding` class in `embedding.py`. These changes improve flexibility in layer initialization, residual connections, and weight normalization. [[1]](diffhunk://#diff-0c893284cd26a98ce6e283e9e05d8caf628110ff759ed10c4c8c4e909b332351R402-R407) [[2]](diffhunk://#diff-0c893284cd26a98ce6e283e9e05d8caf628110ff759ed10c4c8c4e909b332351R419-R426) [[3]](diffhunk://#diff-0c893284cd26a98ce6e283e9e05d8caf628110ff759ed10c4c8c4e909b332351L427-R442) [[4]](diffhunk://#diff-0c893284cd26a98ce6e283e9e05d8caf628110ff759ed10c4c8c4e909b332351R486-R507)

### Flexible Encoder Configurations:
* Made the `encoder` parameter optional in `espnet_model.py`, allowing for cases where no encoder is used. Added support for autocasting in the frontend to improve computational efficiency. [[1]](diffhunk://#diff-cb8db0cbd22ec7ef4211eda6fb02373b3d781248205b935ecf6cfce7e2030c8fL50-R58) [[2]](diffhunk://#diff-cb8db0cbd22ec7ef4211eda6fb02373b3d781248205b935ecf6cfce7e2030c8fL106-R117) [[3]](diffhunk://#diff-cb8db0cbd22ec7ef4211eda6fb02373b3d781248205b935ecf6cfce7e2030c8fL390-R402)

### Layer Normalization Improvements:
* Introduced a new `Dim1LayerNorm` class and its functional equivalent for faster normalization along the first dimension of tensors. This is now an option in CNN-based frontends. [[1]](diffhunk://#diff-7d66d40f560969cd78bf1efaf6f8a9d9ae7a793e7948770f819719a728d932b2L14-R54) [[2]](diffhunk://#diff-7d66d40f560969cd78bf1efaf6f8a9d9ae7a793e7948770f819719a728d932b2R174) [[3]](diffhunk://#diff-7d66d40f560969cd78bf1efaf6f8a9d9ae7a793e7948770f819719a728d932b2R188-R192)

### Other Improvements:
* Updated the convolutional layers in the CNN frontend to use `kaiming_normal_` initialization for better weight initialization.

**Comment by sw005320** [comment]
Thanks, @wanchichen!
